contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java

```diff
@@ -0,0 +1,806 @@
+          );
+    }
+  }
+
+  /** Static class, prevent instantiation. */
+  private KafkaSource() {}
+
+  private static class UnboundedKafkaSource<K, V>
+      extends UnboundedSource<KafkaRecord<K, V>, KafkaCheckpointMark> {
+
+    private final int id; // split id, mainly for debugging
+    private final ImmutableMap<String, Object> consumerConfig;
+    private final List<String> topics;
+    private final Coder<K> keyCoder;
+    private final Coder<V> valueCoder;
+    private final SerializableFunction<KafkaRecord<K, V>, Instant> timestampFn;
+    // would it be a good idea to pass currentTimestamp to watermarkFn?
+    private final Optional<SerializableFunction<KafkaRecord<K, V>, Instant>> watermarkFn;
+    private final List<TopicPartition> assignedPartitions;
+
+    public UnboundedKafkaSource(
+        int id,
+        ImmutableMap<String, Object> consumerConfig,
+        List<String> topics,
+        Coder<K> keyCoder,
+        Coder<V> valueCoder,
+        SerializableFunction<KafkaRecord<K, V>, Instant> timestampFn,
+        Optional<SerializableFunction<KafkaRecord<K, V>, Instant>> watermarkFn,
+        List<TopicPartition> assignedPartitions) {
+
+      this.id = id;
+      this.consumerConfig = consumerConfig;
+      this.topics = topics;
+      this.keyCoder = keyCoder;
+      this.valueCoder = valueCoder;
+      this.timestampFn = timestampFn;
+      this.watermarkFn = watermarkFn;
+      this.assignedPartitions = ImmutableList.copyOf(assignedPartitions);
+    }
+
+    @Override
+    public List<UnboundedKafkaSource<K, V>> generateInitialSplits(
+        int desiredNumSplits, PipelineOptions options) throws Exception {
+
+      KafkaConsumer<K, V> consumer = new KafkaConsumer<K, V>(consumerConfig);
+
+      List<TopicPartition> partitions = Lists.newArrayList();
+
+      // fetch partitions for each topic
+      // sort by <topic, partition>
+      // round-robin assign the partition to splits
+
+      try {
+        for (String topic : topics) {
+          for (PartitionInfo p : consumer.partitionsFor(topic)) {
+            partitions.add(new TopicPartition(p.topic(), p.partition()));
+          }
+        }
+      } finally {
+        consumer.close();
+      }
+
+      Collections.sort(partitions, new Comparator<TopicPartition>() {
+        public int compare(TopicPartition tp1, TopicPartition tp2) {
+          return ComparisonChain
+              .start()
+              .compare(tp1.topic(), tp2.topic())
+              .compare(tp1.partition(), tp2.partition())
+              .result();
+        }
+      });
+
+      Preconditions.checkArgument(desiredNumSplits > 0);
+      Preconditions.checkState(partitions.size() > 0,
+          "Could not find any partitions. Please check Kafka configuration and topic names");
+
+      int numSplits = Math.min(desiredNumSplits, partitions.size());
+
+      List<List<TopicPartition>> assignments = Lists.newArrayList();
```

Comment:
just `new ArrayList<>();` -- the Guava team's recommendation is to stop using these static factory methods since Java 7 has the diamond (`<>`) syntax.